File: Algorithm/ORM_Python/env/sense_obs.py
import numpy as np
from helper.calc_add_limit_vector import calc_vector, limit_angle

def sense_obs(staticObs, movingObs, robot, radSafe):
    """ Calculate the distance and direction from each obstacle to the robot """
    numSObs = staticObs.shape[0]  # 12
    numMObs = movingObs.shape[0]  # 2
    senseObs = np.zeros((numSObs+numMObs, 5),dtype=np.float64)
    posRobot = robot[0:2]
    
    for i in range(0,numSObs):
        posObs = staticObs[i,0:2]
        radObs = staticObs[i,2] + radSafe
        [dirObs, disObs] = list(calc_vector(posRobot, posObs))  
        senseObs[i,:] = [disObs, dirObs, radObs, 0, 0]
        
    for i in range(0,numMObs):
        posObs = movingObs[i,0:2]
        radObs = movingObs[i,3] + radSafe
        vObs = movingObs[i,6]
        [dirObs, disObs] = calc_vector(posRobot, posObs)
        if movingObs[i,7] == -1:
            [dirMove,_] = calc_vector(movingObs[i,4:6], posObs)
            dirMove = limit_angle(dirMove + np.pi/2)
        elif movingObs[i,7] == 0: 
            dirMove = movingObs[i,5]
        elif movingObs[i,7] == 1: 
            [dirMove,_] = calc_vector(movingObs[i,4:6], posObs)
            dirMove = limit_angle(dirMove - np.pi/2)
        senseObs[numSObs+i,:] = [disObs, dirObs, radObs, vObs, dirMove]
    return senseObs

    """ Sort the senseObs from nearest to farest (bubble sort) """
    numObs = numSObs+numMObs
    if numObs > 1:
        for i in range(0,numObs):
            not_swapped = True
            for j in range(1,numObs-i):
                if senseObs[j-1,0] > senseObs[j,0]:
                    temp = senseObs[j-1,:]
                    senseObs[j-1,:] = senseObs[j,:]
                    senseObs[j,:] = temp
                    not_swapped = False
            if not_swapped:
                break
    return senseObs  # 23x5

File: Algorithm/ORM_Matlab/env/sense_obs.m
 function senseObs = sense_obs(staticObs, movingObs, robot, radSafe)
% Calculate the distance and direction from each obstacle to the robot
[numSObs, ~] = size(staticObs);
[numMObs, ~] = size(movingObs);
senseObs = zeros(numSObs+numMObs, 5);
posRobot = robot(1:2);

for i=1:numSObs
    posObs = staticObs(i,1:2);
    radObs = staticObs(i, 3) + radSafe;
    [dirObs, disObs] = calc_vector(posRobot, posObs);
    senseObs(i,:) = [disObs, dirObs, radObs, 0, 0];
end

for i=1:numMObs
    posObs = movingObs(i, 1:2);
    radObs = movingObs(i, 4) + radSafe;
    vObs = movingObs(i,7);
    [dirObs, disObs] = calc_vector(posRobot, posObs);
    switch movingObs(i,8)
        case -1
            [dirMove, ~] = calc_vector(movingObs(i,5:6), posObs);
            dirMove = limit_angle(dirMove + pi/2);
        case 0
            dirMove = movingObs(i,6);
        case 1
            [dirMove, ~] = calc_vector(movingObs(i,5:6), posObs);
            dirMove = limit_angle(dirMove - pi/2);
    end
    senseObs(numSObs+i,:) = [disObs, dirObs, radObs, vObs, dirMove];
end

% Sort the senseObs from nearest to farest (bubble sort)
numObs = numSObs+numMObs;
if numObs > 1
    for i = 0:(numObs-1)
        not_swapped = true;
        for j = 1:(numObs-i-1)
            if senseObs(j,1) > senseObs(j+1,1)
                temp = senseObs(j,:);
                senseObs(j,:) = senseObs(j+1,:);
                senseObs(j+1,:) = temp;
                not_swapped = false;
            end
        end
        if not_swapped
            break
        end
    end
end
    